### PR TITLE
WebXRManager: Added depth sensing support (v2).

### PR DESCRIPTION
**Pull request by cabanier**
Related: #27154.

Simpler approach to depth sensing proposed by @mrdoob.
We copy the depth into the destination frame buffer instead and have it sorted by fl

*This contribution is funded by [Meta](https://meta.com)*

### CODE REVIEW

**Comment by hybridherbst** [comment]
Cool! What's your perception of it?

**Comment by cabanier** [comment]
> Cool! What's your perception of it?

See https://twitter.com/rcabanier/status/1748050448270135425

**Comment by hybridherbst** [comment]
For some reason I get a 90° around Y rotated camera here in our scenes, and frustum culling is wrong. Does this maybe not work when the camera is parented to something or the camera is moving after the session is started?

**Comment by cabanier** [comment]
> For some reason I get a 90° around Y rotated camera here in our scenes, and frustum culling is wrong. Does this maybe not work when the camera is parented to something or the camera is moving after the session is started?

Yeah, maybe that all needs to move to the webxr manager. I'll take a look

**Comment by hybridherbst** [comment]
Thanks!
It feels like performance is quite noticeably better on this version, not sure if you're seeing the same

**Comment by cabanier** [comment]
> Thanks! It feels like performance is quite noticeably better on this version, not sure if you're seeing the same

that's possible. We're not sampling from the surrounding pixels in this version.

**Comment by cabanier** [comment]
> > For some reason I get a 90° around Y rotated camera here in our scenes, and frustum culling is wrong. Does this maybe not work when the camera is parented to something or the camera is moving after the session is started?
> 
> Yeah, maybe that all needs to move to the webxr manager. I'll take a look

@hybridherbst I updated the PR. It's even simpler now

**Comment by hybridherbst** [comment]
Thank you, that works for me now too! Great to be able to compare the two approaches – and a hard decision on what's better suited for three.js 🗡️ 

- v1: less pixelated, lots of potential for artistic exploration of the provided scene depth, technically quite involved
- v2: more pixelated, better performance, not much to adjust, technically very simple

Here's a quick test with custom shaders on v1: https://twitter.com/hybridherbst/status/1748106798173733036

Maybe one possible approach could be:
- v2 is shipped in three.js core and covers the most typical use case
- v1 is provided as an example, using shader patching?

**Comment by cabanier** [comment]
@hybridherbst maybe we can ship @mrdoob 's suggestion as an MVP but also leave the door open for experiences to do better. Maybe we can optionally turn off the automatic occlusion and let experiences define custom shaders to do it themselves.

**Comment by cabanier** [comment]
@mrdoob thoughts?

**Comment by mrdoob** [comment]
Yeah! This feels so much better!

File: src/renderers/WebGLRenderer.js [hunk 1]
```diff
@@ -1209,7 +1209,11 @@ class WebGLRenderer {
 
 			//
 
-			background.render( currentRenderList, scene );
+			if ( xr.enabled === false || xr.isPresenting === false || xr.hasOcclusion() === false ) {
```
Comment (mrdoob) [inline src/renderers/WebGLRenderer.js:1212]: What happens if the browser hasn't implemented `xr.hasOcclusion()`?

**Comment by mrdoob** [comment]
Does this work on Quest 3 already or do I have to enable any flag?

**Comment by Maksims** [comment]
> Does this work on Quest 3 already or do I have to enable any flag?

It works in latest Quest 3 Browser (OS should be updated too I guess), without a need to enable any flags.

**Comment by cabanier** [comment]
> > Does this work on Quest 3 already or do I have to enable any flag?
> 
> It works in latest Quest 3 Browser (OS should be updated too I guess), without a need to enable any flags.

Correct. It's enabled by default. The first time you use it in browser, you should get an OS permission prompt.

**Comment by cabanier** [comment]
> Yeah! This feels so much better!

@mrdoob If you're ok with the changes, we can submit and then fine-tune it so people can still use a custom shader if they want
(ie to get effects like this: https://twitter.com/hybridherbst/status/1748106798173733036)

**Comment by mrdoob** [comment]
I think we should refactor the code into a self contained `WebXRDepthSensing` class.

I'll try to do it over the weekend if you don't beat me to it.

**Comment by cabanier** [comment]
> I'll try to do it over the weekend if you don't beat me to it.

Since you have a solution in mind, it's likely better that you do it :-)
Let me know if you need help

**Comment by cabanier** [comment]
> I think we should refactor the code into a self contained `WebXRDepthSensing` class.
> 
> I'll try to do it over the weekend if you don't beat me to it.

Were you able to get to it? If not, can you describe what you are looking for? I'll try to implement it.

**Comment by mrdoob** [comment]
@cabanier done!

File: src/renderers/webxr/WebXRManager.js [hunk 2]
```diff
@@ -537,6 +550,15 @@ class WebXRManager extends EventDispatcher {
...
 
+				cameraL.near = depthSensing.depthNear; // ?
+				cameraL.far = depthSensing.depthFar; // ?
+				cameraR.near = depthSensing.depthNear; // ?
+				cameraR.far = depthSensing.depthFar; // ?
+
+				cameraL.updateProjectionMatrix();
+				cameraR.updateProjectionMatrix();
+				camera.updateProjectionMatrix();
+
```
Comment (mrdoob) [inline src/renderers/webxr/WebXRManager.js:561]: @cabanier Do you mind explaining these lines? If `depth-sensing` is not available we're setting these to 0, right?

Comment (cabanier) [reply src/renderers/webxr/WebXRManager.js:561]: no, we should leave these to what the author set them or the WebXR default.

Comment (cabanier) [reply src/renderers/webxr/WebXRManager.js:561]: These values come from the author and browser uses them to calculate the matrices

Comment (mrdoob) [reply src/renderers/webxr/WebXRManager.js:561]: So the code looks good to you? If not, feel free to clean up the PR.

Comment (cabanier) [reply src/renderers/webxr/WebXRManager.js:561]: If I submitted this, it is wrong. It should be using _currentDepthNear/_currentDepthFar

**Comment by mrdoob** [comment]
By the way, I feel like the depthTexture is one frame behind (or more).
Are we not fetching it at the right time or is this something that can be improved on the browser side?

**Comment by cabanier** [comment]
> By the way, I feel like the depthTexture is one frame behind (or more). Are we not fetching it at the right time or is this something that can be improved on the browser side?

It should be timewarped to the same time that we use for controller poses and other openxr calls. It's possible though that I'm returning the previous one; I will check.

File: src/renderers/webxr/WebXRManager.js (hunk 2 above)
Comment (mrdoob) [reply src/renderers/webxr/WebXRManager.js:561]: Ok, please change it to the correct values when you get a chance.

Comment (cabanier) [reply src/renderers/webxr/WebXRManager.js]: done

**Comment by mrdoob** [comment]
Okay, I'm going to merge this for now but I'm getting these warnings when inside immersive + depth sensing...

Screenshot 2024-01-31 at 15 33 59

Screenshot 2024-01-31 at 15 39 53

Any ideas?

**Comment by cabanier** [comment]
> Okay, I'm going to merge this for now but I'm getting these warnings when inside immersive + depth sensing...
> Any ideas?

I've seen these as well in regular WebXR sessions. I believe it's because three is using a swapchain texture that is out of band.

**Comment by mrdoob** [comment]
I don't understand what that means...

Are we doing something wrong? Do we have to change something? And if so, what do we need to change? Is it an issue in the Oculus Browser instead?  And if so, are you planning to address it?

**Comment by michaelthatsit** [comment]
I just tried updating and integrating this into mrjs and I'm unable to get it to work.

it's working in the THREE examples just fine. Do I have to use `MeshLambertMaterial` in order for it to work? I'm also playing with renderOrder for our UI Elements, would this have an impact?

I'm also getting those print outs.

**Comment by michaelthatsit** [comment]
Ignore my last comment, we were clearing the renderer which is what caused the issue.

have use cases where we'd like to occlude the hands but selectively occlude planes. Is there a way to adjust the occlusion?

**Comment by cabanier** [comment]
> have use cases where we'd like to occlude the hands but selectively occlude planes. Is there a way to adjust the occlusion?

If you want to occlude on a per object basis, you need to do the occlusion in the shader. Maybe we should give you access to the depth texture and not populate the rendertarget with the depth pixels

**Comment by michaelthatsit** [comment]
that would be great, honestly being able to pass a distance parameter would work too. I'd love to integrate this and abandon our render order hack, but not at the cost of losing a lot of the MR features we've added.

**Comment by michaelthatsit** [comment]
filed a feature request: https://github.com/mrdoob/three.js/issues/27664

**Comment by mrdoob** [comment]
@cabanier can you please answer my question?

**Comment by cabanier** [comment]
I'm unsure what the `glFrameBufferTexture2D` error is. I don't believe that one comes from our side. I'll look into it tomorrow.

**Comment by mrdoob** [comment]
Thank you!

**Comment by cabanier** [comment]
I found and fixed the problem in the browser. It will be part of the next browser release.

**Comment by cabanier** [comment]
@mrdoob I fixed this problem using the WebXR samples. However, it is still happening in three.js.
It doesn't seem to have bad side effects but I'll investigate it a bit more. (For some reason, an internal call is flagging that the previous GL call failed. This is happening every third frame which make it look like a swapchain issue)

**Comment by cabanier** [comment]
> @mrdoob I fixed this problem using the WebXR samples. However, it is still happening in three.js. It doesn't seem to have bad side effects but I'll investigate it a bit more. 

I dug a bit deeper. It seems that drawing the depth texture is generating an invalid operation when it executes drawElements to paint the depth into the framebuffer. Do you know why this could be?

**Comment by cabanier** [comment]
Reverting to fad7c393175f844fbd0069549089b94100022988 solves the problem.

**Comment by cabanier** [comment]
@mrdoob @Mugen87  I figured out that this is happening because we're drawing the depth sensing shader in the shadow map pass. Is there a way to exclude it?

**Comment by odysseyjason** [comment]
Meta Quest 3 on the PTC and received Meta system software v65 on April 26th, following the update none of the Depth Sensing samples perform occlusion any longer.  I do not know if it is a Meta browser issue, threejs issue, or if Meta is no longer providing the browser with access to the depth sensor data on the Meta Quest 3.

I will confirm that none of the following are providing any occlusion now and they previously did work before the Meta v65 update

[threejs webxr_xr_dragging](https://threejs.org/examples/?q=webxr#webxr_xr_dragging)
[playercanvas xr ar-camera depth](https://playcanvas.github.io/#/xr/ar-camera-depth)

(this has me very sad as I find occlusion to be the COOLEST webxr feature to come around!)

**Comment by Mugen87** [comment]
AFAICT, the release versions should not be affected by my changes. There was a temporary issue in `r163dev` but it has been fixed via #27855.

**Comment by cabanier** [comment]
> AFAICT, the release versions should not be affected by my changes. There was a temporary issue in `r163dev` but it has been fixed via #27855.

I didn't look very closely what caused or fixed the issue. depth sensing is broken on the public samples but working on `dev`

**Pull request merged**